src/Common/ZooKeeper/ZooKeeperLoadBalancer.cpp

```diff
@@ -0,0 +1,801 @@
+{
+    switch (load_balancing_type)
+    {
+        case LoadBalancing::RANDOM:
+            return std::make_unique<Random>();
+        case LoadBalancing::NEAREST_HOSTNAME:
+            return std::make_unique<NearestHostname>(getFQDNOrHostName());
+        case LoadBalancing::HOSTNAME_LEVENSHTEIN_DISTANCE:
+            return std::make_unique<Levenshtein>(getFQDNOrHostName());
+        case LoadBalancing::IN_ORDER:
+            return std::make_unique<InOrder>();
+        case LoadBalancing::FIRST_OR_RANDOM:
+            return std::make_unique<FirstOrRandom>();
+        case LoadBalancing::ROUND_ROBIN:
+            return std::make_unique<RoundRobin>();
+    }
+
+    return nullptr;
+}
+
+std::pair<std::string, bool> parseForSocketAddress(const std::string & raw_host)
+{
+    std::pair<std::string, bool> result;
+    bool secure = startsWith(raw_host, "secure://");
+    if (secure)
+        result.first = raw_host.substr(strlen("secure://"));
+    else
+        result.first = raw_host;
+    result.second = secure;
+    return result;
+}
+
+bool isKeeperHostDNSAvailable(Poco::Logger* log, const std::string & address, bool & dns_error_occurred)
+{
+    /// We want to resolve all hosts without DNS cache for keeper connection.
+    Coordination::DNSResolver::instance().removeHostFromCache(address);
+    try
+    {
+        const Poco::Net::SocketAddress host_socket_addr{address};
+    }
+    catch (const Poco::Net::HostNotFoundException & e)
+    {
+        /// Most likely it's misconfiguration and wrong hostname was specified
+        LOG_ERROR(log, "Cannot use ZooKeeper host {}, reason: {}", address, e.displayText());
+        return false;
+    }
+    catch (const Poco::Net::DNSException & e)
+    {
+        /// Most likely DNS is not available now
+        dns_error_occurred = true;
+        LOG_ERROR(log, "Cannot use ZooKeeper host {} due to DNS error: {}", address, e.displayText());
+        return false;
+    }
+    return true;
+}
+
+ZooKeeperLoadBalancer & ZooKeeperLoadBalancer::instance(const std::string & config_name)
+{
+    static std::unordered_map<std::string, ZooKeeperLoadBalancer> load_balancer_by_name;
+    static std::mutex mutex;
+
+    std::lock_guard<std::mutex> lock{mutex};
+    return load_balancer_by_name.emplace(config_name, ZooKeeperLoadBalancer(config_name)).first->second;
+}
+
+ZooKeeperLoadBalancer::ZooKeeperLoadBalancer(const std::string & config_name)
+    : log(&Poco::Logger::get("ZooKeeperLoadBalancer/" + config_name))
+{
+}
+
+void ZooKeeperLoadBalancer::init(zkutil::ZooKeeperArgs args_, std::shared_ptr<ZooKeeperLog> zk_log_)
+{
+    if (args_.hosts.empty())
+        throw zkutil::KeeperException::fromMessage(Coordination::Error::ZBADARGUMENTS, "No hosts specified in ZooKeeperArgs.");
+
+    args = args_;
+    zk_log = std::move(zk_log_);
+
+    connection_balancer = getConnectionBalancer(args.get_priority_load_balancing.load_balancing);
+
+    for (size_t i = 0; i < args_.hosts.size(); ++i)
```

Comment:
@CheSema  In this commit https://github.com/ClickHouse/ClickHouse/pull/56563/commits/ad95bdbf819a96acbe9533299fbfe7e4ea0fd042

I changed it so that 
- `ZooKeeper::startSession` would not re-init the load balancer.
- however, when config reloaded, should be due to `system reload config` after checking code, we will reset;

Additionally as for the question on "do we have concurrent createClient"? I think it could be. We have `auxilary_zookeeper` and can multiple named ZK configuration. https://github.com/ClickHouse/ClickHouse/pull/14155/files#diff-50fdf83595b40f8695c7f6ecdaff323750211e95c8e778f0c8c627912c66115cR17

I think it's better to separate their states with one load balancer instance per config_name. Because while zk host address can be the same, people can configure different timeout/error simulation settings.

After that, mutex within `ZooKeeperLoadBalancer` can be removed.